### PR TITLE
Reduce memory footprint in backupable db.

### PR DESCRIPTION
**Pull request by robertabcd**
For your review. Thanks.

### CODE REVIEW

**Comment by robertabcd** [comment]
Umm, seeing failed test on another try on Travis.

https://travis-ci.org/robertabcd/rocksdb/builds/46322324
db/db_test.cc:9738: Assertion failure Put(Key(count), RandomString(&rnd, 1024), wo).ok()

Not yet reproducible on my machine.
Not sure if the issue is related to this commit.

File: utilities/backupable/backupable_db.cc [hunk 1]
```diff
@@ -145,19 +145,26 @@ class BackupEngineImpl : public BackupEngine {
...
     const uint64_t size;
-    uint32_t checksum_value;
+    const uint32_t checksum_value;
   };
 
   class BackupMeta {
    public:
     BackupMeta(const std::string& meta_filename,
-        std::unordered_map<std::string, FileInfo>* file_infos, Env* env)
+        std::unordered_map<std::string, shared_ptr<FileInfo> >* file_infos,
```
Comment (igorcanadi) [inline utilities/backupable/backupable_db.cc]: can you use std::shared_ptr? Also, no need for a space between > > in C++11 :)

**Comment by igorcanadi** [comment]
Probably unrelated.

Would you also mind producing some performance numbers to demonstrate improved performance?

File: utilities/backupable/backupable_db.cc [hunk 2]
```diff
@@ -207,8 +214,8 @@ class BackupEngineImpl : public BackupEngine {
     uint64_t size_;
     std::string const meta_filename_;
     // files with relative paths (without "/" prefix!!)
-    std::vector<std::string> files_;
-    std::unordered_map<std::string, FileInfo>* file_infos_;
+    std::vector<shared_ptr<FileInfo> > files_;
```
Comment (igorcanadi) [inline utilities/backupable/backupable_db.cc]: Hm, do we even need this structure? We can just use file_infos_ since it contains all the data, right?

**Comment by igorcanadi** [comment]
I do like this patch, though, thanks :)

File: utilities/backupable/backupable_db.cc (hunk 1 above)
Comment (robertabcd) [inline utilities/backupable/backupable_db.cc]: Okay. These will be fixed in the next patch.
btw, unique_ptr is used before my patch. Is this allowed but not shared_ptr?

File: utilities/backupable/backupable_db.cc (hunk 2 above)
Comment (robertabcd) [inline utilities/backupable/backupable_db.cc]: Files may be shared among backups. Each backup is represented by a BackupMeta. file_infos_ contains files mentioned in all meta files. Need a way to store which backup contains which files.

**Comment by robertabcd** [comment]
Hmm. The pull request may be misnamed. The performance gain should be backup engine loading performance.

**Comment by igorcanadi** [comment]
Thanks for contributing!

**Pull request merged**